File: validation/python/validators/validate_handshake_flow.py
from __future__ import annotations

import base64
import hashlib
import hmac
import sys
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[3]
sys.path.append(str(ROOT_DIR / "validation" / "python"))

from util.cbor_canonical import encode_canonical  # type: ignore


def hkdf_sha256(ikm: bytes, info: bytes, length: int = 32) -> bytes:
    salt = b"\x00" * hashlib.sha256().digest_size
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    t1 = hmac.new(prk, info + b"\x01", hashlib.sha256).digest()
    return t1[:length]


def main() -> None:
    path = ROOT_DIR / "tests/common/handshake/end_to_end_test_vectors.json"
    doc = path.read_text(encoding="utf-8")
    vectors = __import__("json").loads(doc)

    flow = vectors.get("handshake_flow") or {}
    steps = flow.get("steps") or []
    if len(steps) < 3:
        raise SystemExit("handshake_flow missing steps")

    resp = steps[1]["message"]
    complete = steps[2]["message"]

    encoded = encode_canonical(resp)
    h = hashlib.sha256(encoded).digest()
    handshake_hash = base64.b64encode(h).decode()
    session_id = base64.b64encode(hkdf_sha256(h, b"FoxWhisper-SessionId", 32)).decode()

    if handshake_hash != complete.get("handshake_hash"):
        raise SystemExit(f"handshake_hash mismatch: {complete.get('handshake_hash')} != {handshake_hash}")
    if session_id != complete.get("session_id"):
        raise SystemExit(f"session_id mismatch: {complete.get('session_id')} != {session_id}")

    print("✅ handshake_flow derivation matches (Python)")


if __name__ == "__main__":
    main()

File: validation/rust/validators/validate_handshake_flow.rs
use base64::{engine::general_purpose, Engine as _};
use hkdf::Hkdf;
use serde::Deserialize;
use serde::ser::{SerializeMap, SerializeSeq};
use serde_cbor::value::{to_value, Value};
use sha2::{Digest, Sha256};
use std::error::Error;
use std::fs;

fn encode_canonical(value: &serde_json::Value) -> Result<Vec<u8>, serde_cbor::Error> {
    let val: Value = to_value(value)?;
    serde_cbor::to_vec(&CanonValue(&val))
}

struct CanonValue<'a>(&'a Value);

impl<'a> serde::Serialize for CanonValue<'a> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self.0 {
            Value::Array(items) => {
                let mut seq = serializer.serialize_seq(Some(items.len()))?;
                for item in items {
                    seq.serialize_element(&CanonValue(item))?;
                }
                seq.end()
            }
            Value::Map(map) => {
                let mut entries = Vec::with_capacity(map.len());
                for (k, v) in map.iter() {
                    let key_bytes = serde_cbor::to_vec(&CanonValue(k)).map_err(serde::ser::Error::custom)?;
                    entries.push((key_bytes, k, v));
                }
                entries.sort_by(|(kb1, _, _), (kb2, _, _)| kb1.len().cmp(&kb2.len()).then_with(|| kb1.cmp(kb2)));

                let mut map_ser = serializer.serialize_map(Some(entries.len()))?;
                for (_, k, v) in entries {
                    map_ser.serialize_entry(&CanonValue(k), &CanonValue(v))?;
                }
                map_ser.end()
            }
            other => other.serialize(serializer),
        }
    }
}

#[derive(Deserialize)]
struct FlowDoc {
    handshake_flow: HandshakeFlow,
}

#[derive(Deserialize)]
struct HandshakeFlow {
    steps: Vec<Step>,
}

#[derive(Deserialize)]
struct Step {
    message: serde_json::Value,
}

fn main() -> Result<(), Box<dyn Error>> {
    let data = fs::read_to_string("tests/common/handshake/end_to_end_test_vectors.json")?;
    let doc: FlowDoc = serde_json::from_str(&data)?;
    let steps = &doc.handshake_flow.steps;
    if steps.len() < 3 {
        return Err("handshake_flow missing steps".into());
    }
    let resp = &steps[1].message;
    let complete = &steps[2].message;

    let encoded = encode_canonical(resp)?;
    let hash = Sha256::digest(&encoded);
    let handshake_hash = general_purpose::STANDARD.encode(hash);

    let hk = Hkdf::<Sha256>::new(None, &hash);
    let mut okm = [0u8; 32];
    hk.expand(b"FoxWhisper-SessionId", &mut okm)
        .map_err(|e| format!("hkdf expand failed: {e}"))?;
    let session_id = general_purpose::STANDARD.encode(okm);

    let expected_hash = complete["handshake_hash"].as_str().unwrap_or("");
    let expected_session = complete["session_id"].as_str().unwrap_or("");

    if handshake_hash != expected_hash {
        return Err(format!("handshake_hash mismatch: {} != {}", expected_hash, handshake_hash).into());
    }
    if session_id != expected_session {
        return Err(format!("session_id mismatch: {} != {}", expected_session, session_id).into());
    }

    println!("✅ handshake_flow derivation matches (Rust)");
    Ok(())
}
